components/sst_importer/src/sst_importer.rs

```diff
@@ -192,80 +232,189 @@ impl SSTImporter {
+            .with_label_values(&["read"])
+            .observe(start_read.saturating_elapsed().as_secs_f64());
+
+        debug!("downloaded file succeed";
+            "name" => src_file_name,
+            "url"  => %url,
+        );
+        Ok(())
+    }
+
+    fn do_download_and_apply<E: KvEngine>(
+        &self,
+        backend: &StorageBackend,
+        name: &str,
+        rewrite_rule: &RewriteRule,
+        cf: &str,
+        speed_limiter: &Limiter,
+        engine: E,
+    ) -> Result<Option<Range>> {
+        let path = self.dir.get_import_path(name)?;
+        self.download_file_from_external_storage(
+            // don't check file lengthn after download file for now.
+            0,
+            name,
+            path.temp.clone(),
+            backend,
+            // don't support encrypt for now.
+            None,
+            speed_limiter,
+        )?;
+        info!("download file finished {}", name);
+
+        // iterator `path.temp` file and performs rewrites and apply.
+        let file = File::open(path.temp)?;
+        let mut reader = BufReader::new(file);
+        let mut buffer = Vec::new();
+        reader.read_to_end(&mut buffer)?;
+
+        let mut event_iter = EventIterator::new(buffer);
+
+        let old_prefix = rewrite_rule.get_old_key_prefix();
+        let new_prefix = rewrite_rule.get_new_key_prefix();
+
+        let perform_rewrite = old_prefix != new_prefix;
+
+        // perform iteration and key rewrite.
+        let mut key = keys::data_key(new_prefix);
+        let new_prefix_data_key_len = key.len();
+        let mut smallest_key = None;
+        let mut largest_key = None;
+
+        while let Some(k) = event_iter.next() {
+            if perform_rewrite {
+                let old_key = &k;
+
+                if !old_key.starts_with(old_prefix) {
+                    return Err(Error::WrongKeyPrefix {
+                        what: "Key in file",
+                        key: old_key.to_vec(),
+                        prefix: old_prefix.to_vec(),
+                    });
+                }
+                key.truncate(new_prefix_data_key_len);
+                key.extend_from_slice(&old_key[old_prefix.len()..]);
+
+                debug!(
+                    "perform rewrite new key: {:?}, new key prefix: {:?}, old key prefix: {:?}",
+                    log_wrappers::Value::key(keys::origin_key(&key)),
+                    log_wrappers::Value::key(new_prefix),
+                    log_wrappers::Value::key(old_prefix),
+                );
+            } else {
+                key = keys::data_key(&k);
+            }
+            let value = Cow::Borrowed(&event_iter.val);
+            // TODO handle delete cf
+            engine.put_cf(cf, &key, &value)?;
+
+            smallest_key = smallest_key.map_or(Some(k.clone()), |sk| {
+                if sk > k { Some(k.clone()) } else { Some(sk) }
+            });
+            largest_key = largest_key.map_or(Some(k.clone()), |lk| {
```

Comment:
I didn't get it.🤔，the first `Some(k.clone())` seems necessary. because `largest_key` lives out of loop.

Reply:
I mean, when the `smallest_key` is `Some(...)`, the evaluation of the first argument of `map_or` (`Some(k.clone())`) would lead to an unnecessary copy.

Maybe just `smallest_key = smallest_key.map_or_else(|_| Some(k.clone()), |sk| { ... })`(which would only copy k once in the second closure)?

Reply:
Okay, Fixed.